File: bigquery/src/main/java/com/dbx/agent/bigquery/BigQueryAgent.java
package com.dbx.agent.bigquery;

import com.dbx.agent.AbstractJdbcAgent;
import com.dbx.agent.ColumnInfo;
import com.dbx.agent.ConnectParams;
import com.dbx.agent.DatabaseInfo;
import com.dbx.agent.ForeignKeyInfo;
import com.dbx.agent.IndexInfo;
import com.dbx.agent.JdbcIdentifiers;
import com.dbx.agent.JsonRpcServer;
import com.dbx.agent.TableInfo;
import com.dbx.agent.TriggerInfo;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public final class BigQueryAgent extends AbstractJdbcAgent {

    @Override
    protected String driverClass() {
        return "com.simba.googlebigquery.jdbc.Driver";
    }

    @Override
    protected String buildJdbcUrl(ConnectParams params) {
        return buildUrl(params);
    }

    @Override
    public List<DatabaseInfo> listDatabases() {
        return unchecked(() -> {
            List<DatabaseInfo> result = new ArrayList<>();
            try (java.sql.Statement stmt = requireConnected().createStatement();
                 ResultSet rs = stmt.executeQuery("SELECT schema_name FROM INFORMATION_SCHEMA.SCHEMATA ORDER BY schema_name")) {
                while (rs.next()) {
                    result.add(new DatabaseInfo(rs.getString(1)));
                }
            }
            return result;
        });
    }

    @Override
    public List<String> listSchemas() {
        List<String> result = new ArrayList<>();
        for (DatabaseInfo database : listDatabases()) {
            result.add(database.getName());
        }
        return result;
    }

    @Override
    public List<TableInfo> listTables(String schema) {
        return unchecked(() -> {
            List<TableInfo> result = new ArrayList<>();
            String quotedSchema = JdbcIdentifiers.INSTANCE.backtick(schema);
            String sql = "SELECT table_name, table_type FROM " + quotedSchema + ".INFORMATION_SCHEMA.TABLES ORDER BY table_name";
            try (java.sql.PreparedStatement stmt = requireConnected().prepareStatement(sql);
                 ResultSet rs = stmt.executeQuery()) {
                while (rs.next()) {
                    result.add(new TableInfo(
                        rs.getString("table_name"),
                        normalizeTableType(rs.getString("table_type")),
                        null
                    ));
                }
            }
            return result;
        });
    }

    @Override
    public List<ColumnInfo> getColumns(String schema, String table) {
        return unchecked(() -> {
            List<ColumnInfo> result = new ArrayList<>();
            String quotedSchema = JdbcIdentifiers.INSTANCE.backtick(schema);
            String sql = "SELECT column_name, data_type, is_nullable " +
                "FROM " + quotedSchema + ".INFORMATION_SCHEMA.COLUMNS " +
                "WHERE table_name = ? " +
                "ORDER BY ordinal_position";
            try (java.sql.PreparedStatement stmt = requireConnected().prepareStatement(sql)) {
                stmt.setString(1, table);
                try (ResultSet rs = stmt.executeQuery()) {
                    while (rs.next()) {
                        result.add(new ColumnInfo(
                            rs.getString("column_name"),
                            rs.getString("data_type"),
                            "YES".equals(rs.getString("is_nullable")),
                            null,
                            false,
                            null,
                            null,
                            null,
                            null,
                            null
                        ));
                    }
                }
            }
            return result;
        });
    }

    @Override
    public List<IndexInfo> listIndexes(String schema, String table) {
        return Collections.emptyList();
    }

    @Override
    public List<ForeignKeyInfo> listForeignKeys(String schema, String table) {
        return Collections.emptyList();
    }

    @Override
    public List<TriggerInfo> listTriggers(String schema, String table) {
        return Collections.emptyList();
    }

    @Override
    public String setSchemaSQL(String schema) {
        return "";
    }

    static String buildUrl(ConnectParams params) {
        String url = "jdbc:bigquery://" + params.getHost() + ":" + params.getPort() + ";ProjectId=" + params.getDatabase();
        String extraParams = trimSemicolons(params.getUrl_params());
        return extraParams.isBlank() ? url : url + ";" + extraParams;
    }

    private static String trimSemicolons(String value) {
        if (value == null) {
            return "";
        }
        return value.trim().replaceAll("^;+", "").replaceAll(";+\\s*$", "");
    }

    private static String normalizeTableType(String type) {
        return "BASE TABLE".equals(type) ? "TABLE" : type;
    }

    public static void main(String[] args) {
        new JsonRpcServer(new BigQueryAgent()).run();
    }
}

File: bigquery/src/main/kotlin/com/dbx/agent/bigquery/BigQueryAgent.kt
package com.dbx.agent.bigquery

import com.dbx.agent.*
import java.sql.Connection
import java.sql.DriverManager
import java.sql.ResultSet
import java.sql.Types

class BigQueryAgent : DatabaseAgent {
    private var connection: Connection? = null

    override fun getConnection(): Connection? = connection

    override fun connect(params: ConnectParams) {
        Class.forName("com.simba.googlebigquery.jdbc.Driver")
        val url = "jdbc:bigquery://${params.host}:${params.port};ProjectId=${params.database}"
        connection = DriverManager.getConnection(url, params.username, params.password)
    }

    override fun testConnection(params: ConnectParams): Boolean {
        Class.forName("com.simba.googlebigquery.jdbc.Driver")
        val url = "jdbc:bigquery://${params.host}:${params.port};ProjectId=${params.database}"
        DriverManager.getConnection(url, params.username, params.password).use { conn ->
            return conn.isValid(5)
        }
    }

    override fun listDatabases(): List<DatabaseInfo> {
        val conn = connection ?: throw IllegalStateException("Not connected")
        val result = mutableListOf<DatabaseInfo>()
        conn.createStatement().use { stmt ->
            stmt.executeQuery("SELECT schema_name FROM INFORMATION_SCHEMA.SCHEMATA ORDER BY schema_name").use { rs ->
                while (rs.next()) {
                    result.add(DatabaseInfo(rs.getString(1)))
                }
            }
        }
        return result
    }

    override fun listSchemas(): List<String> {
        return listDatabases().map { it.name }
    }

    override fun listTables(schema: String): List<TableInfo> {
        val conn = connection ?: throw IllegalStateException("Not connected")
        val result = mutableListOf<TableInfo>()
        val quotedSchema = JdbcIdentifiers.backtick(schema)
        conn.prepareStatement(
            "SELECT table_name, table_type FROM $quotedSchema.INFORMATION_SCHEMA.TABLES ORDER BY table_name"
        ).use { stmt ->
            stmt.executeQuery().use { rs ->
                while (rs.next()) {
                    val tableType = rs.getString("table_type").let {
                        if (it == "BASE TABLE") "TABLE" else it
                    }
                    result.add(TableInfo(rs.getString("table_name"), tableType))
                }
            }
        }
        return result
    }

    override fun getColumns(schema: String, table: String): List<ColumnInfo> {
        val conn = connection ?: throw IllegalStateException("Not connected")
        val result = mutableListOf<ColumnInfo>()
        val quotedSchema = JdbcIdentifiers.backtick(schema)
        conn.prepareStatement(
            """
            SELECT column_name, data_type, is_nullable
            FROM $quotedSchema.INFORMATION_SCHEMA.COLUMNS
            WHERE table_name = ?
            ORDER BY ordinal_position
            """.trimIndent()
        ).use { stmt ->
            stmt.setString(1, table)
            stmt.executeQuery().use { rs ->
                while (rs.next()) {
                    result.add(
                        ColumnInfo(
                            name = rs.getString("column_name"),
                            data_type = rs.getString("data_type"),
                            is_nullable = rs.getString("is_nullable") == "YES",
                            column_default = null,
                            is_primary_key = false,
                            extra = null,
                            comment = null,
                            numeric_precision = null,
                            numeric_scale = null,
                            character_maximum_length = null
                        )
                    )
                }
            }
        }
        return result
    }

    override fun listIndexes(schema: String, table: String): List<IndexInfo> {
        // BigQuery does not support indexes
        return emptyList()
    }

    override fun listForeignKeys(schema: String, table: String): List<ForeignKeyInfo> {
        // BigQuery does not support foreign keys
        return emptyList()
    }

    override fun listTriggers(schema: String, table: String): List<TriggerInfo> {
        // BigQuery does not support triggers
        return emptyList()
    }

    override fun executeQuery(sql: String, schema: String?, options: ExecuteQueryOptions): QueryResult {
        return JdbcExecutor.execute(
            conn = connection ?: throw IllegalStateException("Not connected"),
            sql = sql,
            schema = schema,
            setSchemaSql = ::setSchemaSQL,
            maxRows = options.maxRows,
            fetchSize = options.fetchSize,
            valueReader = ::getResultValue,
        )
    }

    override fun setSchemaSQL(schema: String): String = ""

    override fun disconnect() {
        connection?.close()
        connection = null
    }

    private fun getResultValue(rs: ResultSet, index: Int, sqlType: Int): Any? {
        val value = when (sqlType) {
            Types.BIGINT -> rs.getLong(index)
            Types.INTEGER, Types.SMALLINT, Types.TINYINT -> rs.getInt(index)
            Types.FLOAT, Types.REAL -> rs.getFloat(index)
            Types.DOUBLE -> rs.getDouble(index)
            Types.DECIMAL, Types.NUMERIC -> rs.getBigDecimal(index)
            Types.BOOLEAN, Types.BIT -> rs.getBoolean(index)
            else -> rs.getString(index)
        }
        return if (rs.wasNull()) null else value
    }
}

fun main() {
    JsonRpcServer(BigQueryAgent()).run()
}
